lib/rucio/core/transfer.py

```diff
@@ -723,379 +1067,205 @@ def get_transfer_requests_and_source_replicas(total_workers=0, worker_number=0,
-        except NoDistance:
-            logger(logging.WARNING, "Request %s: no link from %s to %s", req_id, source_rse_name, dest_rse_name)
-            if req_id in reqs_scheme_mismatch:
-                reqs_scheme_mismatch.remove(req_id)
-            if req_id not in reqs_no_source:
-                reqs_no_source.append(req_id)
-            continue
-        except RSEProtocolNotSupported:
-            logger(logging.WARNING, "Request %s: no matching protocol between %s and %s", req_id, source_rse_name, dest_rse_name)
-            if req_id in reqs_no_source:
-                reqs_no_source.remove(req_id)
-            if req_id not in reqs_scheme_mismatch:
-                reqs_scheme_mismatch.append(req_id)
-            continue
-
-        source_scheme = list_hops[0]['source_scheme']
-        dest_scheme = list_hops[-1]['dest_scheme']
-        dest_scheme_priority = list_hops[-1]['dest_scheme_priority']
+        candidate_paths = []
+        for source in rws.sources:
+            if allowed_source_rses is not None and source.rse_id not in allowed_source_rses:
+                continue
 
-        allow_tape_source = True
-        try:
-            # Get source protocol
-            source_protocol = ctx.protocol(source_rse_id, source_scheme, 'read')
-
-            source_sign_url = ctx.rse_attrs(source_rse_id).get('sign_url', None)
-            dest_sign_url = ctx.rse_attrs(dest_rse_id).get('sign_url', None)
-
-            # Compute the source URL
-            source_url = list(source_protocol.lfns2pfns(lfns={'scope': scope.external, 'name': name, 'path': path}).values())[0]
-            source_url = __rewrite_source_url(source_url, source_sign_url=source_sign_url, dest_sign_url=dest_sign_url, source_scheme=source_scheme)
-
-            # If the request_id is not already in the transfer dictionary, need to compute the destination URL
-            if req_id not in transfers:
-
-                # parse allow tape source expression, not finally version.
-                # allow_tape_source = attr["allow_tape_source"] if (attr and "allow_tape_source" in attr) else True
-                allow_tape_source = True
-
-                # Extend the metadata dictionary with request attributes
-                transfer_src_type = "DISK"
-                transfer_dst_type = "DISK"
-                overwrite, bring_online = True, None
-                if ctx.is_tape_rse(source_rse_id) or ctx.rse_attrs(source_rse_id).get('staging_required', False):
-                    bring_online = bring_online_local
-                    transfer_src_type = "TAPE"
-                    if not allow_tape_source:
-                        if req_id not in reqs_only_tape_source:
-                            reqs_only_tape_source.append(req_id)
-                        if req_id in reqs_no_source:
-                            reqs_no_source.remove(req_id)
-                        continue
+            # rse_name will be None if rse is staging area
```

Comment:
Where is this enforced?

Reply:
Didn't think this through. The comment was there before the refactoring. Do you feel like it is wrong ?